pyam/plotting.py

```diff
@@ -419,19 +424,99 @@ def stack_plot(df, x='year', y='value', stack='variable',
+        for i, cross in enumerate(crosses):
+            cross = cross[0]  # get location
+            x_1 = time_original[cross]
+            x_2 = time_original[cross + 1]
+            y_1 = values[cross]
+            y_2 = values[cross + 1]
+
+            zero_time = x_1 - y_1 * (x_2 - x_1) / (y_2 - y_1)
+            if i == 0:
+                first_zero_times.loc[:, col] = zero_time
+            if zero_time not in _df.index.values:
+                _df.loc[zero_time, :] = np.nan
+
+    first_zero_times = first_zero_times.sort_values(
+        by="first_zero_time",
+        axis=1,
+    )
+    _df = _df.reindex(sorted(_df.index)).interpolate(method="values")
+
+    # Sort lines so that negative timeseries are on the right, positive
+    # timeseries are on the left and timeseries which go from positive to
+    # negative are ordered such that the timeseries which goes negative first
+    # is on the right (case of timeseries which go from negative to positive
+    # is an edge case we haven't thought about as it's unlikely to apply to
+    # us).
+    col_order = (
+        [c for c in _df if (_df[c] > 0).all()]
+        + first_zero_times.columns[::-1].tolist()
+        + [c for c in _df if (_df[c] < 0).all()]
+    )
+    _df = _df[col_order]
+
     # explicitly get colors
     defaults = default_props(reset=True, num_colors=len(_df.columns),
                              colormap=cmap)['color']
     rc = run_control()
-    colors = []
+    colors = {}
     for key in _df.columns:
         c = next(defaults)
-        if 'color' in rc and stack in rc['color'] and\
-                key in rc['color'][stack]:
+        c_in_rc = 'color' in rc
+        if c_in_rc and stack in rc['color'] and key in rc['color'][stack]:
             c = rc['color'][stack][key]
-        colors.append(c)
-
-    ax.stackplot(_df.index, _df.T, labels=_df.columns, colors=colors, **kwargs)
+        colors[key] = c
+
+    # plot stacks, starting from the top and working our way up to the bottom
+    negative_only_cumulative = _df.applymap(
+        lambda x: x if x < 0 else 0
+    ).cumsum(axis=1)
+    positive_only_cumulative = _df.applymap(lambda x: x if x >= 0 else 0)[
+        col_order[::-1]
+    ].cumsum(axis=1)[
+        col_order
+    ]
+    time = _df.index.values
+    upper = positive_only_cumulative.iloc[:, 0].values
+    for j, col in enumerate(_df):
+        noc_tr = negative_only_cumulative.iloc[:, j].values
+        try:
+            poc_nr = positive_only_cumulative.iloc[:, j + 1].values
+        except IndexError:
+            poc_nr = np.zeros_like(upper)
+        lower = poc_nr.copy()
+        if (noc_tr < 0).any():
+            lower[np.where(poc_nr == 0)] = noc_tr[np.where(poc_nr == 0)]
+
+        ax.fill_between(time, lower, upper, label=col,
+                        color=colors[col], **kwargs)
+        upper = lower.copy()
+
+    # add total
+    if total is not None:
+        if not isinstance(total, dict):
```

Comment:
Here, we should be a bit more generic, e.g.,

```
from collections import Mapping

if not isinstance(total, Mapping):
```

Reply:
I guess unless the `Mapping` interface does not support `setdefault()`, then we should probably raise a `TypeError`